### ISSUE TITLE
Control ML workflow via `config.seabee.yaml`

### ISSUE THREAD

**Issue by JamesSample** [SeaBee-no/documentation#54]
Ideally, users should be able to control the ML workflow via `config.seabee.yaml`, in a similar way to how they can currently control settings for ODM.

I know you've already done some work on automating the detections pipeline, @knl88. I haven't looked at this in detail yet, but I will soon.

@knl88 @27am what do you think of the following suggestions:

1. Add a new mandatory Boolean to the config. file called `classify`. If `True`, this will trigger automatic classification of the orthophoto using NR's ML models.
2. Add new attributes `ml_options` (just like `odm_options` in the current specification [here](https://seabee-no.github.io/documentation/data-upload.html#sec-config-file)). These will be mandatory if `classify=True`. As a minimum, I guess we need to specify the type and version of the ML model (which should link to the annotation version), plus a `publish_result` Boolean to determine whether the classified output gets published to GeoNode. Any other options required to configure the NR container should also be included here.
3. Add a new module to `seabeepy` for ML-specific code (`seabeepy.ml` ?). Add a function that accepts the `ml_options` read from the config. file and then does whatever is necessary to trigger the ML workflow (e.g. writes a settings file somewhere in the correct format). 
4. Extend `flights_runner.ipynb` to use the new function(s) to trigger the ML workflow wherever it is specified in the config file. Also add sections to the notebook that check for any completed ML tasks, or any errors produced by the ML workflow.
5. Add a function to `seabeepy.ml` that converts results to a vector layer and publishes to GeoNode (if `publish_result=True` in `ml_options`. The styling of the vector layer should use the colours defined in the annotation scheme for the relevant ML model.
6. Extend  `flights_runner.ipynb` to use the new function(s) for publishing results.

Does this sound reasonable?

**Comment by knl88** [comment]
Nice, think this fits fine into the stuff I looked it, and is a nice step forward:) A couple of thoughts below

>     3. Add a new module to `seabeepy` for ML-specific code (`seabeepy.ml` ?). Add a function that accepts the `ml_options` read from the config. file and then does whatever is necessary to trigger the ML workflow (e.g. writes a settings file somewhere in the correct format).
Made a couple of functions in https://github.com/SeaBee-no/seabeepy/blob/main/notebooks/detections_runner.ipynb that could go in here. It uses the seabee-ai programs, the slightly more elegant approach is to use seabee-ai as a library. 
>     4. Extend `flights_runner.ipynb` to use the new function(s) to trigger the ML workflow wherever it is specified in the config file. Also add sections to the notebook that check for any completed ML tasks, or any errors produced by the ML workflow.
Would perhaps keep them separate, then we have the opportunity to let the ml workflow keep the gpu, although as a start they would probably be executed in the same pod.
>     5. Add a function to `seabeepy.ml` that converts results to a vector layer and publishes to GeoNode (if `publish_result=True` in `ml_options`. The styling of the vector layer should use the colours defined in the annotation scheme for the relevant ML model.
As a start we can perhaps use the publishing of geopackages in seabeepy once we have the vector layer. I also made a PR on geoserver-rest so in the new release we could consider using [create_gpkg_datastore](https://github.com/gicait/geoserver-rest/blob/45544688349f8310215a6d9a0260beb4d10b9731/geo/Geoserver.py#L1825).  A possible alternative is to store the results in a postgres table, but haven't tried publishing through geoserver for this.
> 
>     7. Update mission config. files with the new information. E.g. all NIVA missions can be set to `classify=True`. Probably most/all seabirds missions too?
We would need the gpu when we run all the missions also perhaps check if NR have some new models and perhaps a new slightly more memory efficient release ready. 
> 
> 
> Does this sound reasonable?

Yep:)

We also need to think about the folder structure for the ml results.  In my playground I setup a  [results](https://minio.seabee.sigma2.no/buckets/hub-test-bucket/browse/dGVsZW1hcmtfdHZpc3RlaW5meXJfMjAyMzA1MTUv) folder under each mission folder. The structure of the results folder is based on the output from seabee-ai .i.e `TASK/MODEL_NAME/RUN_DATETIME/output`  where output is this case is either csv or gpkg.

**Comment by knl88** [comment]
> 2. Add new attributes ml_options (just like odm_options in the current specification [here](https://seabee-no.github.io/documentation/data-upload.html#sec-config-file)). These will be mandatory if classify=True. As a minimum, I guess we need to specify the type and version of the ML model (which should link to the annotation version), plus a publish_result Boolean to determine whether the classified output gets published to GeoNode. Any other options required to configure the NR container should also be included here.

`odm_options` are optional should we try to make `ml_options` optional also and provide some defaults?

**Comment by JamesSample** [comment]
Yeah, that sounds good. I guess defining a default model might be a bit fiddly, as we'll need to specify it based on the `theme` attribute in the config. file (i.e. whether it's seabirds, habitats or mammals). I don't think this should be too hard, but if it's a pain we can try something simpler.
